File: solutions/erlang/meetup/1/src/meetup.erl
-module(meetup).

-export([meetup/4]).


meetup(Year, Month, DayOfWeek, Week) -> 
	LDayOfWeek=proplists:lookup_all(
					day_to_num(DayOfWeek),
						for(
							calendar:day_of_the_week(Year, Month,1),
							1,
							calendar:last_day_of_the_month(Year, Month),
							[]
						)
					),
	{Year, Month, get_day(LDayOfWeek,Week)}.

for(_WStart,DayStart,DayEnd,Acc) when DayStart>DayEnd->lists:reverse(Acc);
for(WStart,DayStart,DayEnd,Acc) when WStart>7->for(1,DayStart,DayEnd,Acc);
for(WStart,DayStart,DayEnd,Acc)->for(WStart+1,DayStart+1,DayEnd,[{WStart,DayStart}|Acc]).

get_day(L,first)->{_,Day}=lists:nth(1,L),Day;
get_day(L,second)->{_,Day}=lists:nth(2,L),Day;
get_day(L,third)->{_,Day}=lists:nth(3,L),Day;
get_day(L,fourth)->{_,Day}=lists:nth(4,L),Day;
get_day(L,last)->{_,Day}=lists:last(L),Day;
get_day(L,teenth)->{value,{_,Day}}=lists:search(fun({_,D}) when D >=13->true;(_)->false end, L),Day.

day_to_num(monday)->1;
day_to_num(tuesday)->2;
day_to_num(wednesday)->3;
day_to_num(thursday)->4;
day_to_num(friday)->5;
day_to_num(saturday)->6;
day_to_num(sunday)->7.

File: solutions/javascript/meetup/1/meetup.js
/**
 * @param {number} year
 * @param {number} month
 * @param {string} which
 * @param {string} weekday
 */
export const meetup = (year, month, which, weekday) => {
    const lastDayOfMonth = new Date(year, month, 0);
    const daysInMonth = lastDayOfMonth.getDate();

    /** @type {Record<string, number>} */
    const weekdayIndex = {
        Sunday: 0,
        Monday: 1,
        Tuesday: 2,
        Wednesday: 3,
        Thursday: 4,
        Friday: 5,
        Saturday: 6,
    };
    const targetWeekdayNum = weekdayIndex[weekday];

    const occurrences = [];
    for (let day = 1; day <= daysInMonth; day++) {
        const date = new Date(year, month - 1, day);
        if (date.getDay() === targetWeekdayNum) {
            occurrences.push(date);
        }
    }

    switch (which) {
        case 'first':
            return occurrences[0];
        case 'second':
            return occurrences[1];
        case 'third':
            return occurrences[2];
        case 'fourth':
            return occurrences[3];
        case 'last':
            return occurrences[occurrences.length - 1];
        case 'teenth':
            return occurrences.find(
                (date) => date.getDate() >= 13 && date.getDate() <= 19,
            );
    }

    throw new Error('Invalid parameters');
};
